### DISCUSSION TITLE
Reversing an open Position by matching the reversing entry order type rather than just using market order to exit

### DISCUSSION THREAD

**Discussion by lavalampmj** [MicroTrendsLtd/NinjaTrader8#102] [Q&A] [answered]
The way the NT managed order manager reverses an open position is to match the reversing order type to exit the open position first; for example:

I'm in a long position with 5 contracts.  
My strategy issues a EnterShortLimit(1, myprice) order.
The NT managed order manager will, under the covers,  first place a Sell Limit 5 with myprice and SellShort Limit 1 with myprice. 

Given they were entered milliseconds apart, these orders will live close to each other on the FIFO order book and therefore both be likely to be executed within milliseconds of each other. The benefit of this approach is that exit pricing and entry prices are the same but more importantly the open position is gracefully exited at the same time and only if the price matches the new intrabar entry order type. In other words the reversal (or exit) occurs if the entry order is hit.

This order manager currently uses market orders to immediately exit the open position.  This maybe a considered design choice IDK.  

However, I'm wondering how difficult it would be to emulate the NT managed approach?  

Perhaps to have a property such as, IsMatchReversingEntryOrderType so this manager's behavior can emulate the NT manager's or retain the current design.

**Comment by lavalampmj** [comment]
Would this enhancement just require an update to the "public virtual void PositionClose(bool isPositionCloseModeLimit = false)"
method to check what the entryOrder type is, use it's stopmarket and/or limit prices as applicable and then issue the matching exit order type?

**Comment by MicroTrendsTom** [comment]
Ok so the problem with limits is they don't always get filled so we used market to ensure a fill and this makes the whole thing more robust and simpler.   

Limit exits could be implemented 

Caveats:
It might hit a margin rejection with too many orders open etc.
- but what of the limits that do not get filled - only part filled etc...  
- it would stay - there could be a limit order expiry period and - fill or kill approach.

So there would be a can of worms to deal with... but yes its on the face of it simple to do.
But for my schedule i have 0 time left in 2023 to look at this project currently

so it could be an additional forked version to start if you wanted to try etc?

**Comment by kconcannon** [comment]
Limit exits are a really bad idea.  If you’re had limits yesterday on NQ
you probably would have had a very bad day.

**Comment by lavalampmj** [comment]
Actually I was not suggesting just Limit Exit per se.  My suggestion was to emulate how the NT managed approach behaves today when reversing and existing an existing Position with any opposing enter order type. Letting the user decide whether to use markets order exclusively (as today).

The NT managed reversal approach is robust and simple from a strategy developers standpoint. But obviously NT Managed falls seriously short elsewhere, hence I'm here...

NT managed, under the covers, simply issues another matching order of the same type (MIT, Stop or Limit etc) against the existing position, just prior to issuing the same type entry order.  These two orders, (depending on the type) live close to each other on the exchange order book and would therefore more likely be affected similarly based on current market dynamics, i.e. get hit and filled within milliseconds of each other or not at all.

I'm not 100% certain on this next point, but I think there would be zero incremental margin penalty for issuing an exit order for an open position (as the margin is already accounted when the entry was made)?

Of course, there are caveats with part fills, expiry etc, but those are also the applicable for any intrabar entry orders too, limit and stop limit specifically. 

After looking at the code today, it would not be a simply matter of changing the exit order in close the PositionClose method, as you really need two paths, one to service AlgoSignal.ExitTrade / ExitTradeLong and ExitTradeShort and the other to service AlgoSignal.GoLong / GoShort and GoOCO in the case where an intrabar entry type is used and there is an existing position.   Honestly, this is a rather complex piece of code and I'm reticent to change it given the possible consequences....

**Reply by MicroTrendsTom** [reply]
it not possible for me to say if it can be done or not or the complexity or impact as i don't understand what the use case is etc.
it could just be another workflow needs to be added for a certain mode of trading to handle orders etc.  CAR/SAR/OCO etc

**Comment by MicroTrendsTom** [comment]
NT by default will close at market - cancel and then place the reversal orders - stop,stoplimit or limit or market you define
unless you are exiting and reversing with a  different order type .

Margin
So if you were long 4 with 4 exits and sent a cancel reversal at same time for 4 it can  - say thats 8 lots margin flagged until the cancels are done.   Also if you are 4 long and issue a cancel, short 8  and place exits 4 that is also double margin for the flight and the reduced to 4 when working.

So this is why its good practice to cancel wait then reverse or to have an account size that can handle double the margin + for your system

"Reversing an open Position by matching the reversing entry order type rather than just using market order to exit"
Not sure i follow a position does not have a working order and could be there by any number of events and sequences and not as a result of one specific entry order last fired..     so it could be a random sequence to be repeated as an exit - there would not be any trading advantage in such a system and it would be complex and unreliable.    i must be misinterpreting what you mean.

Much better to state up front to work with 
simple rules of the algo - when an exit occurs and when a trade occurs etc
what is the order entry type
what is the exit order type
what is the exit model CAR,SAR etc
fill or kill or cancel  etc
reverse only if flat etc

Exits and Entry
Would this revolve around different techniques for reversal?

CAR  - close and reverse  -  usually a cancel flatten  -  this can then use any order type for the entry as per exchange rules etc.
stop,stop limit, market, limit etc
SAR - stop and reverse    -  usually a market order stop loss for the stop and additional a stop at market for the reversal and this can be a different price -   stop limits can be gapped over and also don't always get filled so its not a robust approach on a faster system but ok on  a slower provided you put in processing for part fills - do nothing or fill or kill is the usual format. 
SAR could also trigger a limit entry on fill to be filled immediately or later on a pullback - but that would be unusual. 

Trading systems User requirements
To understand what is required  would be easier with some trade examples - stages with the orders and rules in play to understand it
annotated charts and rules etc.

**Comment by lavalampmj** [comment]
Attached is a simple (aka hack) test strategy that shows the reversal behavior under NT's managed approach when reversing an open position with an opposing direction Entry order with LIMIT, MIT and StopMarket.  Apply to low timeframe second chart.  In each case you can see that the NT order manager matches the order type to exit the position prior to the order being placed.

[testreverse.txt](https://github.com/MicroTrendsLtd/NinjaTrader8/files/11466674/testreverse.txt)

Given this order manager uses signals to start flows, prior to orders being placed I am unsure how it could be adapted to support this behavior as you don't know what the order type is until its placed. I guess the solution could be to call a wrapper method around the order placement instead of directly calling SubmitUnmanaged, so you could inspect the order type before placing it?  IDK.

**Reply by MicroTrendsTom** [reply, marked as answer]
ok its easier to break it down into rules and the real order types  and flow .

What i am seeing is Entry and reversal combinations using order types sequences: 
Limit And Reverse (LAR)
Stop And Reverse (SAR)
Market If Touched And Reverse (MTAR)
Close At Market  (CAM)

you can break the orders into multiple or merge them into one for the exit and reverse

Algo Trade Workflow 1
Market Long -  LAR Short -  Close
1)When entry true -   Long  5@ market   - 
2)When  Exit and Reverse true -      -   Short 6 @ limit  (LAR)
3) When Exit short true -    Long 1 @ Market 

Algo Trade Workflow 2
Market Long -  SAR Short -  Close

1)When entry true -   Long  5 @ market   - 
2)When  Exit Stop And Reverse true -      -   Short 6 @ Stop Market   (SAR)
3) When Exit short  true -    Long 1 @ Market 

Algo Trade Workflow 3
Market Long -  MTAR Short -  Close

1)When entry true -   Long  5@ market   - 
2)When Exit And Reverse true -      -   Short 6 @ MIT  (MTAR)
3) When Exit short true -    Long 1 @ Market 

Thats what i can make out you want to do roughly



Yes correct this not supported and overriding code as far as i can think wouldnt do it.
But the following are all easily enough added as discreet workflows - into the trade workflow et
SAR 
LAR
MTAR

Then a strategy can easily do this
1) enter and do not place exits - work to bypass default exit orders
2)trigger the SAR/LAR/MTAR - workflow needed
3) exit condition at market - no work to be done

SAR was planned at some point
they are all basically the same thing with different order types
but no ETA due to a different project obligation.

i think your fastest easiest way would be a custom made unmanaged mode strategy to do the rules
those rules above could  be done very quickly.
